### PR TITLE
Fix/pub sub

### LINKED ISSUE #100: `RedisPubSub.fromConnection` is broken due to wrapping of all connection types in `TimeoutConnection`
`RedisPubSub.fromConnection` tries to pattern match on the connection type, but always fails because all connections are wrapped in a `TimeoutConnection`.

### LINKED ISSUE #104: Stack overflow in `RedisResult[List[(K, V)]]`
Looks like https://github.com/davenverse/rediculous/blob/main/core/shared/src/main/scala/io/chrisdavenport/rediculous/RedisResult.scala#L125-L144 isn't stack safe and can fail for large lists (this happened to me).

### PR DESCRIPTION
**Pull request by buntec**
Attempts to fix #100 and #104.

### CODE REVIEW

**Comment by lenguyenthanh** [comment]
Thanks, I have the same issue. And this fixed the issue for me.

File: core/shared/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala [hunk 1]
```diff
@@ -671,10 +671,4 @@ object RedisConnection{
   private def raceNThrowFirst[F[_]: Concurrent, A](nel: NonEmptyList[F[A]]): F[A] = 
     Stream(Stream.emits(nel.toList).evalMap(identity)).covary[F].parJoinUnbounded.take(1).compile.lastOrError
 
-  private class TimeoutConnection[F[_]: Temporal](rC: RedisConnection[F], duration: Duration) extends RedisConnection[F] {
-
-    def runRequest(inputs: Chunk[NonEmptyList[ByteVector]], key: Option[ByteVector]): F[Chunk[Resp]] =
-      rC.runRequest(inputs, key).timeout(duration)
```
Comment (armanbilge) [inline core/shared/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala:677]: Sorry if I missed it. What replaces this `.timeout(...)` call? I don't think I saw any new `.timeout(...)` calls introduced above.

Comment (buntec) [reply core/shared/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala:677]: Oops, you are right, will try to fix it.

Comment (lenguyenthanh) [reply core/shared/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala:677]: I took a quick look, and have a naive solution, which is add `timeout` to `request` function of each `Connection` class. Would it work?


```diff
M core/shared/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala
@@ -40,7 +40,7 @@ object RedisConnection{
       }
     } 
   }
-  private[rediculous] case class PooledConnection[F[_]: Concurrent](
+  private[rediculous] case class PooledConnection[F[_]: Temporal](
     pool: KeyPool[F, Unit, Socket[F]], timeout: Duration
   ) extends RedisConnection[F]{
     def runRequest(inputs: Chunk[NonEmptyList[ByteVector]], key: Option[ByteVector]): F[Chunk[Resp]] = {
@@ -52,6 +52,7 @@ object RedisConnection{
           case _ => Applicative[F].unit
         }
       }.rethrow
+      .timeout(timeout)
     }
   }
 
```

Comment (buntec) [reply core/shared/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala:677]: Yes, that's what I've done now. I've also sealed the `RedisConnection` trait to ensure exhaustive matches in `RedisPubSub`. Or is that going too far?

Comment (lenguyenthanh) [reply core/shared/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala:677]: > I've also sealed the `RedisConnection` trait to ensure exhaustive matches in `RedisPubSub`. Or is that going too far?

I think that's a good idea.

**Review by armanbilge — COMMENTED** [review]
Wouldn't a more straightforward fix be to keep `TimeoutConnection`, add it to the pattern match, and handle that case recursively?

I think adding `sealed` is good, but it's a source-breaking change that probably Chris should sign-off on.

If I understand correctly: for the record, this sort of bug would be prevented with proper linting enabled?

**Comment by buntec** [comment]
Agree, that seems better!

**Comment by lenguyenthanh** [comment]
> If I understand correctly: for the record, this sort of bug would be prevented with proper linting enabled?

I would believe so.

**Review by armanbilge — APPROVED** [review]

**Pull request merged**